Source/JavaScriptCore/wasm/WasmTypeDefinition.h

```diff
@@ -469,7 +469,7 @@ inline void Type::dump(PrintStream& out) const
     TypeKind kindToPrint = kind;
     if (index != TypeDefinition::invalidIndex) {
         auto signedIndex = static_cast<std::make_signed<TypeIndex>::type>(index);
-        if (signedIndex < 0) {
+        if ((signedIndex < 0) && (signedIndex >= minTypeValue)) {
```

Comment:
Why is it different in 32bit? I think this is fundamental invariant, and having difference in 32bit sounds wrong, and the right fix is fixing 32bit not to have this wrong condition.
Can you explain the reason of 32bit difference?

Reply:
(sorry for barging in but timezone differences may make this more expedient) as I understand, the problem is that on 64-bit unix/linux (and even windows, apparently); user-space pointers are never going to have the sign bit set, so we can directly pun them as `intptr_t` and be assured the result is going to be positive

but on 32-bit linux, there's no such restriction: user-space pointers can, in principle, have the MSB set [1], so we need something different as a test for "is this a TypeDefinition* or a TypeKind"

... though I agree fixing the linked bug would be a more robust solution

[1] https://linux-kernel-labs.github.io/refs/heads/master/lectures/address-space.html#linux-address-space

Reply:
On top of the address space split being configurable, it is very much the case that userland pointers may have the MSB set in the 32-bit-userland-running-on-a-64-bit-kernel used by on the ARM buildbots (necessary to take advantage of more powerful ARM hardware). In fact, userspace gets a separate address space: http://sprunge.us/zpMowI. Not sure if they're doing that to mitigate spectre or for some other reason.

Reply:
That makes sense. So I think the right fix is that we should use first 0-4KB pointers for these types (probably not using 0 since it should be used for null / invalid). The first page for address 0-4KB is ensured to be unmapped in C language standard, and all modern OSes ensure this (otherwise, nullptr can point at some memory). So these 0-4KB pointers can be used reliably.
Right now, TypeKind is using negative values. But we should just use small integer values, and then, we can check this on 64bit and 32bit platforms.

```
if (value < 4KB)
    this is TypeKind
```
I think that removing 32bit specific hack is super important, and in this case, I think this is the right fix.

Reply:
I understand that type indices being invalid pointers is desirable. However, the current values of type indices are not arbitrary -- they're the [binary representation](https://www.w3.org/TR/wasm-core-1/#types%E2%91%A6) of the corresponding wasm types.

I see two paths forward:

1. Flip the sign of the runtime representation of wasm types when loading them. This feels like introducing more opportunity for errors.
2. Use the C++ type system to avoid errors in the interpretation of the type index (https://bugs.webkit.org/show_bug.cgi?id=247454)

Option 2 looks like the clear winner here, but maybe you have some better idea.

In any case, this feels like a "nice to have" on 32-bits. The type index representation is still invalid as a pointer on 64-bits, it's only 32-bits that doesn't have this property. With that in mind, I'd rather land this simple fix now to keep the tests green and solve this by option 2 above at a later stage, what do you think?

Reply:
I think it is OK to use the current simple check. But right now, this patch is only changing it in `Type::dump`, is it the only place?
Also, if 32bit has this kind of problem, then `(signedIndex < 0) && (signedIndex >= minTypeValue)` check should be extracted as a function.

Reply:
OK, did that and updated the PR. AFAICT this is the only site where the code checks for < 0.

Reply:
Ping.